src/Nonlinear/SymbolicAD/SymbolicAD.jl

```diff
@@ -0,0 +1,225 @@
+# Copyright (c) 2017: Miles Lubin and contributors
+# Copyright (c) 2017: Google Inc.
+#
+# Use of this source code is governed by an MIT-style license that can be found
+# in the LICENSE.md file or at https://opensource.org/licenses/MIT.
+
+module SymbolicAD
+
+import MathOptInterface as MOI
+
+"""
+    simplify(f)
+
+Return a simplified version of the function `f`.
+
+!!! warning
+    This function is not type stable by design.
+"""
+simplify(f) = f
+
+function simplify(f::MOI.ScalarAffineFunction{T}) where {T}
+    f = MOI.Utilities.canonical(f)
+    if isempty(f.terms)
+        return f.constant
+    end
+    return f
+end
+
+function simplify(f::MOI.ScalarQuadraticFunction{T}) where {T}
+    f = MOI.Utilities.canonical(f)
+    if isempty(f.quadratic_terms)
+        return simplify(MOI.ScalarAffineFunction(f.affine_terms, f.constant))
```

Comment:
Calling `simplify` will canonicalize again, we don't need to do the type unstable part of deciding whether it should be a constant.

Reply:
I don't understand this comment

Reply:
we just did `f = MOI.Utilities.canonical(f)` so `f.affine_terms` are already canonicalized. Here, we create a SAF that we know is canonical and then pass it to `simplify`. The first thing the function will do is canonicalize again which is a bit wasteful